### DISCUSSION TITLE
How to upload file？

### DISCUSSION THREAD

**Discussion by kitepad** [c-smile/sciter-js-sdk#94] [Q&A]
My scenario:  
1. User  open a FolderSelection dialog.
2. Select single or multiple file (.png .jpg, .gif or .docx)
3. upload selected files to remote server.

Should I use Fetch API or mithril.js. I don't know if mithril.js  FileUpload/XHR is workable with Sciter.js.

**Comment by c-smile** [comment]
Fetch API supports file upload, I'll make sample for that.

**Comment by kitepad** [comment]
Andrew, thanks for your help.  Expect to see your sample soon.

**Comment by c-smile** [comment]
See  [samples/fetch(AJAX)/test-upload.htm](https://github.com/c-smile/sciter-js-sdk/blob/main/samples/fetch(AJAX)/test-upload.htm) in 4.4.7.4

**Comment by kitepad** [comment]
I tried this sample and it's running ok. But I don't understand this statement and I cann't find any document about it.
`    let dataToSend = {"file1": file, "comment1": "this is just a test" };

    var resp = await fetch("https://sciter.com/test/test-file-receiver.php", {
      method: "POST",
      body: dataToSend
    });`

file1 and comment1 are used in dataToSend as json key.  Is it  neededy by demo php file? 
I have to write my server logic to parse body and I cann't issue standard form submission request. I find some javascript demo code to upload file , but most of them required FormData, but is not exist in sciter.js.

**Comment by kitepad** [comment]
I write a sample server using Deno (https://deno.land) : 
`import Aqua from "https://deno.land/x/aqua/mod.ts";
const app = new Aqua(3100);
app.get("/", (req) => {
    return "Hello, World!";
});
app.post("/upload", async (req) => {
    console.log(req.files);
    return "Uploaded!";
});`.

When I use curl to upload file, success.  curl -XPOST -F "file=@test-fetch.htm" http://127.0.0.1:3100/upload.
But it's failed when run scapp test-upload.htm. Only change fetch's URL to   http://127.0.0.1:3100/upload.

**Comment by c-smile** [comment]
This in Sciter:
```JS
let dataToSend = {"file1": file, "comment1": "this is just a test" };

var resp = await fetch("https://sciter.com/test/test-file-receiver.php", {
  method: "POST",
  body: dataToSend
});
```
is equivalent of 

```JS
    const dataToSend = new FormData();
    dataToSend.append('file1',file);    
    dataToSend.append('comment1',"this is just a test");    
    const resp = await fetch("https://sciter.com/test/test-file-receiver.php", {
          method:'POST',
          body:dataToSend
     } );
```
in browsers.     

If body is an object (map) that contains File objects, Sciter will send `Content-Type: multipart/form-data`  data, otherwise `Content-Type: application/x-www-form-urlencoded` -  in the same way as browsers do.

**Comment by kitepad** [comment]
I tried your sample code, I think Content-Type: multipart/form-data should be sent, but I see is "content-type": "application/x-www-form-urlencoded;"

**Reply by kitepad** [reply]
I found， I cann't change content-type, even I follow fetch api spec to set header, my added new header can add into request header.

**Reply by kitepad** [reply]
Below is my code:
`    let file = await sys.fs.open(URL.toPath(__DIR__ + "test-upload.htm"));
    let dataToSend = {"file1": file,"comment1":"test"}; 
    var resp = await fetch("http://127.0.0.1:3100/upload", {
     method: "POST",
      headers:{"content-type":"miltipart/form-data",
        "x-test-header":"text/plain"},
      body: dataToSend
   });`

**Reply by kitepad** [reply]
How can I change content-type header and others？

**Reply by c-smile** [reply]
Could you explain why do you need to change content-type header?

If that's about posting data then check: https://stackoverflow.com/a/26791188/421163

**Reply by kitepad** [reply]
Base on your comment : "If body is an object (map) that contains File objects, Sciter will send Content-Type: multipart/form-data data, otherwise Content-Type: application/x-www-form-urlencoded - in the same way as browsers do."

I think Sciter should send multipart/form-data content-type for test-upload.htm, but my demo server received content-type is  application/x-www-form-urlencoded. I think it's not correct. 
Below is my server log for your reference: 

  url: "/upload",
  headers: {
    accept: "*/*",
    "accept-encoding": "gzip, deflate",
    **"content-type": "application/x-www-form-urlencoded;charset=utf-8",**
    "user-agent": "Sciter.JS/4.4.7.3 (Windows;Windows-10.2;en)",
    host: "127.0.0.1:3100",
    "content-length": "0",
    connection: "Keep-Alive",
    "cache-control": "no-cache"
  },
  method: "POST",

**Reply by kitepad** [reply]
I check my demo server using curl to upload same file, server log output is content-type: multipart/form-data. I don't know what's happened and why.

**Reply by kitepad** [reply]
I know why. It is my mistake. It's ok after I upgrade to 4.4.7.4. 

Thanks.
